File: src/bpp_shared/world/noise/noise_simplex.cpp
/*
 * Copyright (c) 2025-2026, Pixel Brush <pixelbrush.dev>
 *
 * SPDX-License-Identifier: AGPL-3.0-only
 * Based on code by Mojang Studios (2011)
*/

#include "noise_simplex.h"
#include <array>
#include <cassert>

static constexpr std::array<uint8_t, 256> MOD12 = [] {
	std::array<uint8_t, 256> table{};
	for (size_t i = 0; i < table.size(); ++i)
		table[i] = uint8_t(i % 12);
	return table;
}();

NoiseSimplex::NoiseSimplex() {
	Java::Random rand;
	InitPermTable(rand);
}

NoiseSimplex::NoiseSimplex(Java::Random& rand) {
	InitPermTable(rand);
}

void NoiseSimplex::InitPermTable(Java::Random& rand) {
	coordinate.x = rand.nextDouble() * 256.0;
	coordinate.y = rand.nextDouble() * 256.0;
	coordinate.z = rand.nextDouble() * 256.0;

	for (int32_t i = 0; i < 256; ++i) {
		permutations[i] = i;
	}

	for (int32_t i = 0; i < 256; ++i) {
		int32_t j = rand.nextInt(256 - i) + i;
		std::swap(permutations[i], permutations[j]);
		permutations[i + 256] = permutations[i];
	}
}

void NoiseSimplex::GenerateNoise(std::vector<double>& values, Vec2 p_offset, Int32_2 p_size, Vec2 p_scale,
                                 double amplitude, bool overwrite) {
	size_t index = 0;

	for (int32_t xI = 0; xI < p_size.x; ++xI) {
		double xPos = (p_offset.x + double(xI)) * p_scale.x + coordinate.x;

		for (int32_t yI = 0; yI < p_size.y; ++yI) {
			double yPos = (p_offset.y + double(yI)) * p_scale.y + coordinate.y;
			double skew = (xPos + yPos) * skewing;
			int32_t x0 = wrap(xPos + skew);
			int32_t y0 = wrap(yPos + skew);
			double unskewed = double(x0 + y0) * unskewing;
			double x0a = double(x0) - unskewed;
			double y0a = double(y0) - unskewed;
			double x0b = xPos - x0a;
			double y0b = yPos - y0a;
			int8_t i;
			int8_t j;
			if (x0b > y0b) {
				i = 1;
				j = 0;
			} else {
				i = 0;
				j = 1;
			}

			double x0c = x0b - double(i) + unskewing;
			double y0c = y0b - double(j) + unskewing;
			double x1c = x0b - 1.0 + 2.0 * unskewing;
			double y1c = y0b - 1.0 + 2.0 * unskewing;
			int32_t xInt = x0 & 255;
			int32_t yInt = y0 & 255;
			int32_t grad0 = MOD12[size_t(permutations[xInt + permutations[yInt]])];
			int32_t grad1 = MOD12[size_t(permutations[xInt + i + permutations[yInt + j]])];
			int32_t grad2 = MOD12[size_t(permutations[xInt + 1 + permutations[yInt + 1]])];
			double term0 = 0.5 - x0b * x0b - y0b * y0b;
			double contrib0;
			if (term0 < 0.0) {
				contrib0 = 0.0;
			} else {
				term0 *= term0;
				contrib0 = term0 * term0 * dotProd(gradients[grad0], x0b, y0b);
			}

			double term1 = 0.5 - x0c * x0c - y0c * y0c;
			double contrib1;
			if (term1 < 0.0) {
				contrib1 = 0.0;
			} else {
				term1 *= term1;
				contrib1 = term1 * term1 * dotProd(gradients[grad1], x0c, y0c);
			}

			double term2 = 0.5 - x1c * x1c - y1c * y1c;
			double contrib2;
			if (term2 < 0.0) {
				contrib2 = 0.0;
			} else {
				term2 *= term2;
				contrib2 = term2 * term2 * dotProd(gradients[grad2], x1c, y1c);
			}

			if (overwrite)
				values[index++] = 70.0 * (contrib0 + contrib1 + contrib2) * amplitude;
			else
				values[index++] += 70.0 * (contrib0 + contrib1 + contrib2) * amplitude;
		}
	}
}

File: crates/bpp_shared/src/world/noise/noise_simplex.rs
/*
 * Copyright (c) 2025-2026, Pixel Brush <pixelbrush.dev>
 *
 * SPDX-License-Identifier: AGPL-3.0-only
 * Based on code by Mojang Studios (2011)
*/

// A recreation of the the Infdev 20100227-1433 Perlin noise function
use crate::helpers::java::java_math::double_to_int32;
use crate::helpers::java::java_random::Random;
use crate::numeric_structs::{Int32_2, VEC3_ZERO, Vec2, Vec3};

/// @brief A faithful reimplementation of the Beta-era simplex noise generator, often used for Biome generation
pub struct NoiseSimplex {
    pub permutations: [i32; 512],
    pub coordinate: Vec3,
    gradients: [[i32; 3]; 12],
    skewing: f64,
    unskewing: f64,
}

impl NoiseSimplex {
    pub fn new() -> Self {
        let mut rand = Random::new();
        let mut noise = Self::blank();
        noise.init_perm_table(&mut rand);
        noise
    }

    pub fn with_random(rand: &mut Random) -> Self {
        let mut noise = Self::blank();
        noise.init_perm_table(rand);
        noise
    }

    fn blank() -> Self {
        Self {
            permutations: [0; 512],
            coordinate: VEC3_ZERO,
            gradients: [
                [1, 1, 0],
                [-1, 1, 0],
                [1, -1, 0],
                [-1, -1, 0],
                [1, 0, 1],
                [-1, 0, 1],
                [1, 0, -1],
                [-1, 0, -1],
                [0, 1, 1],
                [0, -1, 1],
                [0, 1, -1],
                [0, -1, -1],
            ],
            skewing: 0.5 * (3.0f64.sqrt() - 1.0),
            unskewing: (3.0 - 3.0f64.sqrt()) / 6.0,
        }
    }

    fn init_perm_table(&mut self, rand: &mut Random) {
        self.coordinate.x = rand.next_double() * 256.0;
        self.coordinate.y = rand.next_double() * 256.0;
        self.coordinate.z = rand.next_double() * 256.0;

        for i in 0..256usize {
            self.permutations[i] = i as i32;
        }

        for i in 0..256usize {
            let j = (rand.next_int_bound(256 - i as i32) + i as i32) as usize;
            self.permutations.swap(i, j);
            self.permutations[i + 256] = self.permutations[i];
        }
    }

    pub fn generate_noise(&self, values: &mut Vec<f64>, p_offset: Vec2, p_size: Int32_2, p_scale: Vec2, amplitude: f64) {
        let mut index = 0usize;

        for x_i in 0..p_size.x {
            let x_pos = (p_offset.x + x_i as f64) * p_scale.x + self.coordinate.x;

            for y_i in 0..p_size.y {
                let y_pos = (p_offset.y + y_i as f64) * p_scale.y + self.coordinate.y;
                let skew = (x_pos + y_pos) * self.skewing;
                let x0 = wrap(x_pos + skew);
                let y0 = wrap(y_pos + skew);
                let unskewed = (x0 + y0) as f64 * self.unskewing;
                let x0a = x0 as f64 - unskewed;
                let y0a = y0 as f64 - unskewed;
                let x0b = x_pos - x0a;
                let y0b = y_pos - y0a;
                let i: i8;
                let j: i8;
                if x0b > y0b {
                    i = 1;
                    j = 0;
                } else {
                    i = 0;
                    j = 1;
                }

                let x0c = x0b - i as f64 + self.unskewing;
                let y0c = y0b - j as f64 + self.unskewing;
                let x1c = x0b - 1.0 + 2.0 * self.unskewing;
                let y1c = y0b - 1.0 + 2.0 * self.unskewing;
                let x_int = (x0 & 255) as usize;
                let y_int = (y0 & 255) as usize;
                let grad0 = self.permutations[x_int + self.permutations[y_int] as usize] % 12;
                let grad1 = self.permutations[x_int + i as usize + self.permutations[y_int + j as usize] as usize] % 12;
                let grad2 = self.permutations[x_int + 1 + self.permutations[y_int + 1] as usize] % 12;
                let mut term0 = 0.5 - x0b * x0b - y0b * y0b;
                let contrib0 = if term0 < 0.0 {
                    0.0
                } else {
                    term0 *= term0;
                    term0 * term0 * dot_prod(&self.gradients[grad0 as usize], x0b, y0b)
                };

                let mut term1 = 0.5 - x0c * x0c - y0c * y0c;
                let contrib1 = if term1 < 0.0 {
                    0.0
                } else {
                    term1 *= term1;
                    term1 * term1 * dot_prod(&self.gradients[grad1 as usize], x0c, y0c)
                };

                let mut term2 = 0.5 - x1c * x1c - y1c * y1c;
                let contrib2 = if term2 < 0.0 {
                    0.0
                } else {
                    term2 *= term2;
                    term2 * term2 * dot_prod(&self.gradients[grad2 as usize], x1c, y1c)
                };

                values[index] += 70.0 * (contrib0 + contrib1 + contrib2) * amplitude;
                index += 1;
            }
        }
    }
}

impl Default for NoiseSimplex {
    fn default() -> Self {
        Self::new()
    }
}

pub fn wrap(grad: f64) -> i32 {
    if grad > 0.0 { double_to_int32(grad) } else { double_to_int32(grad) - 1 }
}

pub fn dot_prod(grad: &[i32; 3], x: f64, y: f64) -> f64 {
    grad[0] as f64 * x + grad[1] as f64 * y
}
